### PR TITLE
additional asserts and comments in torrent::on_blocks_hashed

### PR DESCRIPTION
**Pull request by arvidn**
This is me trying to understand the constraints on this function. As far as I can tell, these additional asserts should hold.

### CODE REVIEW

File: src/torrent.cpp [hunk 1]
```diff
@@ -4040,22 +4040,45 @@ namespace {
...
 				break;
 			}
-			auto result = get_hash_picker().set_block_hash(piece, i * default_block_size, block_hashes[i]);
+			auto const result = get_hash_picker().set_block_hash(piece
+				, i * default_block_size, block_hashes[i]);
 
-			block_passed[i] = result.status == set_block_hash_result::result::success;
 			if (result.status == set_block_hash_result::result::success)
 			{
-				TORRENT_ASSERT(result.first_verified_block <= blocks_in_piece);
```
Comment (arvidn) [inline src/torrent.cpp:4048]: my understanding is that `first_verified_block` should never be `== blocks_in_piece`. That would only make sense if `num_verified == 0`. I strengthened this assert by saying it's `< blocks_in_piece`

File: src/torrent.cpp [hunk 2]
```diff
@@ -4040,22 +4040,45 @@ namespace {
...
+				TORRENT_ASSERT(result.first_verified_block < blocks_in_piece);
+				TORRENT_ASSERT(blocks_in_piece <= blocks_per_piece);
+
+				// all verified ranges should always be full pieces or less
+				TORRENT_ASSERT(result.first_verified_block >= 0
+					|| (result.first_verified_block % blocks_per_piece) == 0);
+				TORRENT_ASSERT(result.num_verified < blocks_in_piece
+					|| (result.num_verified % blocks_per_piece) == 0);
```
Comment (arvidn) [inline src/torrent.cpp:4055]: it seems to hold true, that if more than one piece was validated, we only get whole pieces.

File: src/torrent.cpp [hunk 3]
```diff
@@ -4040,22 +4040,45 @@ namespace {
...
+				// first_verified_block may be negative.
+
+				// In this call, we track the blocks in this piece though, in
+				// the block_passed array. For that tracking we need to clamp
+				// the start index to 0.
 				auto const first_block = std::max(0, result.first_verified_block);
-				std::fill_n(block_passed.begin() + first_block
-					, std::min(blocks_in_piece - first_block, result.num_verified), true);
+				auto const count = std::min(blocks_in_piece - first_block, result.num_verified);
+				std::fill_n(block_passed.begin() + first_block, count, true);
```
Comment (arvidn) [inline src/torrent.cpp:4068]: the only change here is factoring out of the `count`, just to be able to use it in the assert below

File: src/torrent.cpp [hunk 4]
```diff
@@ -4040,22 +4040,45 @@ namespace {
 				ret = false;
 				break;
 			}
-			auto result = get_hash_picker().set_block_hash(piece, i * default_block_size, block_hashes[i]);
+			auto const result = get_hash_picker().set_block_hash(piece
+				, i * default_block_size, block_hashes[i]);
 
-			block_passed[i] = result.status == set_block_hash_result::result::success;
```
Comment (arvidn) [inline src/torrent.cpp:4045]: when we allocate `block_passed`, we clear it to all be `false`. Setting it to false here should not be necessary. What's left is setting it to `true`, which we can do unconditionally in the if-block below. But, in fact, we already do a `fill_n()` to set fields to `true` inside the if-block, and the range we set should always include `i`, which means we don't actually need this assignment anymore.

File: src/torrent.cpp [hunk 5]
```diff
@@ -4040,22 +4040,45 @@ namespace {
...
 
-				// if the hashes for more than one piece have been verified, check for any pieces which
-				// were already checked but couldn't be verified and mark them as verified
-				for (piece_index_t verified_piece = piece + piece_delta(result.first_verified_block / blocks_per_piece)
-					, end = verified_piece + piece_delta(result.num_verified / blocks_per_piece)
+				// if the hashes for more than one piece have been verified,
+				// check for any pieces which were already checked but couldn't
+				// be verified and mark them as verified
+				for (piece_index_t verified_piece = piece + delta(result.first_verified_block / blocks_per_piece)
+					, end = verified_piece + delta(result.num_verified / blocks_per_piece)
```
Comment (arvidn) [inline src/torrent.cpp:4081]: it's asserted above that `first_verified_block` and `num_verified` should be divisible by `blocks_per_piece` (unless we verified less than a full piece, in which case this loop isn't relevant anyway)

**Comment by arvidn** [comment]
@ssiloti does my understanding seem correct?

File: src/torrent.cpp (hunk 4 above)
Comment (ssiloti) [reply src/torrent.cpp:4045]: Right. IIRC the verified block range and corresponding fill_n came much later than this line. I clearly didn't notice that the former obsoleted the latter.

**Pull request merged**